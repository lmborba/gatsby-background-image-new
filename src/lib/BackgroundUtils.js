/**
 * Gets styles by a class name.
 *
 * @notice The className has to exactly match the CSS class
 * @param className string
 */
import { isBrowser, isString, stringToArray, toCamelCase } from './SimpleUtils';

/**
 * Gets styles rules by a class name.
 *
 * @notice The className has to exactly match the CSS class
 * @param className string
 */
export const getStyleRulesForClassName = className => {
  const styleSheets = isBrowser() ? window.document.styleSheets : [];
  for (let i = 0; i < styleSheets.length; i++) {
    let classes;
    try {
      classes =
        typeof styleSheets[i].rules !== 'undefined'
          ? styleSheets[i].rules
          : typeof styleSheets[i].cssRules !== 'undefined'
          ? styleSheets[i].cssRules
          : '';
    } catch (e) {}
    if (classes) {
      const foundClass = Array.prototype.slice
        .call(classes)
        .reduce((foundAcc, styleRule) => {
          if (styleRule.selectorText === className) {
            foundAcc.push(styleRule);
          }
          return foundAcc;
        }, []);
      if (foundClass.length) {
        return foundClass;
      }
    }
  }
  return [];
};

/**
 * Fixes non-enumerable style rules in Firefox.
 *
 * @param cssStyleRules CSSStyleRules   DOM-StyleRules-Object
 * @return {*}
 */
export const getStyleRules = cssStyleRules => {
  let styles = {};
  if (
    cssStyleRules.length > 0 &&
    typeof cssStyleRules[0].style !== 'undefined'
  ) {
    // Fallback for Browsers without constructor.name (IE11).
    const constructorName =
      cssStyleRules[0].style.constructor.name ||
      cssStyleRules[0].style.constructor.toString();

    switch (constructorName) {
      // For Firefox or IE11.
      case 'CSS2Properties':
      case '[object MSStyleCSSProperties]':
        Object.values(cssStyleRules[0].style).forEach(prop => {
          styles[toCamelCase(prop)] = cssStyleRules[0].style[prop];
        });
        break;
      case 'CSSStyleDeclaration':
        styles = cssStyleRules[0].style;
        break;
      default:
        console.error('Unknown style object prototype');
        break;
    }
  }
  return styles;
};

/**
 * Filters out Background Rules for a given class Name.
 *
 * @param className   string    The class to filter rules from
 * @return {{}}
 */
export const getBackgroundStylesForSingleClass = className => {
  if (className && isString(className)) {
    const cssStyleRules = getStyleRulesForClassName(`.${className}`);
    // const cssStyleRules = rulesForCssText(style)

    if (
      cssStyleRules?.length > 0 &&
      typeof cssStyleRules[0].style !== 'undefined'
    ) {
      // Filter out background(-*) rules that contain a definition.
      return Object.keys(getStyleRules(cssStyleRules))
        .filter(
          key =>
            key.indexOf('background') === 0 &&
            cssStyleRules[0].style[key] !== ''
        )
        .reduce((newData, key) => {
          newData[toCamelCase(key)] = cssStyleRules[0].style[key];
          return newData;
        }, {});
    }
  }
  return {};
};

/**
 * Uses the above to get all background(-*) rules from given class(es).
 *
 * @param className   string|array    className or array of classNames
 * @return {*}
 */
const getBackgroundStyles = className => {
  //if (isBrowser()) {
    const classes = stringToArray(className);
    if (classes instanceof Array) {
      const classObjects = [];
      classes.forEach(item =>
        classObjects.push(getBackgroundStylesForSingleClass(item))
      );
      return Object.assign(...classObjects);
    }
    return getBackgroundStylesForSingleClass(className);
  //}
  //return {};
};

export default getBackgroundStyles;
