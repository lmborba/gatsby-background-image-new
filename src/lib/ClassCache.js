const componentClassCache = Object.create({});
/**
 * Cache component classes as we never know if a Component wasn't already set.
 *
 * @param className   string  className given by props
 * @return {*|boolean}
 */
export const inComponentClassCache = className => {
  return componentClassCache[className] || false;
};

/**
 * Adds a component's classes to componentClassCache.
 *
 * @param className   string  className given by props
 */
export const activateCacheForComponentClass = className => {
  if (className) {
    componentClassCache[className] = true;
  }
};

/**
 * Resets the componentClassCache (especially important for reliable tests).
 */
export const resetComponentClassCache = () => {
  for (const className in componentClassCache)
    delete componentClassCache[className];
};
