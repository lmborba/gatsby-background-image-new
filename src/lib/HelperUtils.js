import filterInvalidDOMProps from 'filter-invalid-dom-props';
import { groupByMedia, hasArtDirectionSupport } from './MediaUtils';

/**
 * Strip BackgroundImage propTypes from remaining props to be passed to <Tag />
 *
 * @param props
 * @return {Object}
 */
export const stripRemainingProps = props => filterInvalidDOMProps(props);

/**
 * Preprocess art directed images.
 *
 * @param props
 * @return {Object}
 */
export const convertProps = props => {
  const convertedProps = { ...props };
  const { fixed, fluid } = convertedProps;

  // if (fluid && !hasImageArray(props)) {
  //   convertedProps.fluid = [].concat(fluid)
  // }
  //
  // if (fixed && !hasImageArray(props)) {
  //   convertedProps.fixed = [].concat(fixed)
  // }

  // convert fluid & fixed to arrays so we only have to work with arrays
  if (fluid && hasArtDirectionSupport(props, 'fluid')) {
    convertedProps.fluid = groupByMedia(convertedProps.fluid);
  }
  if (fixed && hasArtDirectionSupport(props, 'fixed')) {
    convertedProps.fixed = groupByMedia(convertedProps.fixed);
  }

  return convertedProps;
};
