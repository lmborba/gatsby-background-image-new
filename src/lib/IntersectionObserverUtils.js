import { isBrowser } from "./SimpleUtils";

let io;
const listeners = new WeakMap();

/**
 * Executes each IntersectionObserver entries' callback.
 *
 * @param entries
 */
export const callbackIO = entries => {
  entries.forEach(entry => {
    if (listeners.has(entry.target)) {
      const callback = listeners.get(entry.target);
      // Edge doesn't currently support isIntersecting, so also test for an intersectionRatio > 0
      if (entry.isIntersecting || entry.intersectionRatio > 0) {
        io.unobserve(entry.target);
        listeners.delete(entry.target);
        callback();
      }
    }
  });
};

/**
 * Returns an IntersectionObserver instance if exists.
 *
 * @param rootMargin    string    The current rootMargin, defaults to 200px.
 * @return {IntersectionObserver|undefined}
 */
export const getIO = rootMargin => {
  if (
    typeof io === `undefined` &&
    isBrowser() &&
    window.IntersectionObserver
  ) {
    io = new window.IntersectionObserver(callbackIO, {
      rootMargin,
    });
  }

  return io;
};

/**
 * Registers IntersectionObserver callback on element.
 *
 * @param element     Element   The element to observe.
 * @param callback    function  Callback to call when intersecting.
 * @param rootMargin  string    The current rootMargin, defaults to 200px.
 * @return {Function}
 */
export const listenToIntersections = (
  element,
  callback,
  rootMargin = `200px`
) => {
  const observer = getIO(rootMargin);

  if (observer) {
    observer.observe(element);
    listeners.set(element, callback);
    return () => {
      observer.unobserve(element);
      listeners.delete(element);
    };
  }
  return () => {};
};
