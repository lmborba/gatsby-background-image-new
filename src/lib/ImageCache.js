import { convertProps } from './HelperUtils';
import { hasArtDirectionArray } from './MediaUtils';
import { getImageSrcKey, getSelectedImage, hasImageArray } from './ImageUtils';
import { isObject } from './SimpleUtils';

const imageCache = Object.create({});
/**
 * Cache if we've seen an image before so we don't bother with
 * lazy-loading & fading in on subsequent mounts.
 *
 * @param props
 * @param index
 * @param isLoop
 * @return {boolean}
 */
export const inImageCache = (props, index = 0, isLoop = false) => {
  const convertedProps = convertProps(props);
  const isImageStack =
    hasImageArray(convertedProps) && !hasArtDirectionArray(convertedProps);
  if (isImageStack && !isLoop) {
    return allInImageCache(props);
  }

  // Find src
  const src = isImageStack
    ? getSelectedImage(convertedProps, index)
    : getImageSrcKey(convertedProps);
  if (isObject(src)) {
    const objectSrc = getImageSrcKey({ fluid: src, fixed: src });
    return imageCache[objectSrc] || false;
  }
  return imageCache[src] || false;
};

/**
 * Processes an image stack with inImageCache.
 *
 * @param props  object    Component Props (with fluid or fixed as array).
 * @return {*|boolean}
 */
export const allInImageCache = props => {
  const convertedProps = convertProps(props);
  // Extract Image Array.
  const imageStack = convertedProps.fluid || convertedProps.fixed;
  // Only return true if every image is in cache.
  return imageStack.every((imageData, index) => {
    return inImageCache(convertedProps, index, true);
  });
};

/**
 * Adds an Image to imageCache.
 *
 * @param props
 * @param index
 * @param isLoop
 */
export const activateCacheForImage = (props, index = 0, isLoop = false) => {
  const convertedProps = convertProps(props);
  const isImageStack =
    hasImageArray(convertedProps) && !hasArtDirectionArray(convertedProps);
  if (isImageStack && !isLoop) {
    return activateCacheForMultipleImages(props);
  }
  // Find src
  const src = isImageStack
    ? getSelectedImage(convertedProps, index)
    : getImageSrcKey(convertedProps);
  if (src) {
    if (isObject(src)) {
      const objectSrc = getImageSrcKey({ fluid: src, fixed: src });
      imageCache[objectSrc] = true;
    } else {
      imageCache[src] = true;
    }
  }
};

/**
 * Activates the Cache for multiple Images.
 *
 * @param props
 */
export const activateCacheForMultipleImages = props => {
  const convertedProps = convertProps(props);
  // Extract Image Array.
  const imageStack = convertedProps.fluid || convertedProps.fixed;
  imageStack.forEach((imageData, index) =>
    activateCacheForImage(convertedProps, index, true)
  );
};

/**
 * Resets the image cache (especially important for reliable tests).
 */
export const resetImageCache = () => {
  for (const prop in imageCache) delete imageCache[prop];
};
