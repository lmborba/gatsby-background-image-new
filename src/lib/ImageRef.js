import { convertProps } from './HelperUtils';
import {
  getCurrentSrcData,
  getSelectedImage,
  hasImageArray,
  hasPictureElement,
  imageLoaded,
} from './ImageUtils';
import {
  createArtDirectionSources,
  createSourceElementForSrcSet,
  hasArtDirectionArray,
} from './MediaUtils';
import { isBrowser, isString } from './SimpleUtils';
import { inImageCache } from './ImageCache';

/**
 * Creates an image reference to be activated on critical or visibility.
 * @param props
 * @param onLoad
 * @param index
 * @param isLoop
 * @return {HTMLImageElement|null|Array}
 */
export const createPictureRef = (props, onLoad, index, isLoop = false) => {
  const convertedProps = convertProps(props);

  if (
    isBrowser() &&
    (typeof convertedProps.fluid !== `undefined` ||
      typeof convertedProps.fixed !== `undefined`)
  ) {
    const isImageStack =
      hasImageArray(convertedProps) && !hasArtDirectionArray(convertedProps);
    if (isImageStack && !isLoop) {
      return createMultiplePictureRefs(props, onLoad);
    }
    const img = new Image();

    img.onload = () => onLoad();

    if (!img.complete && typeof convertedProps.onLoad === `function`) {
      img.addEventListener('load', convertedProps.onLoad);
    }
    if (typeof convertedProps.onError === `function`) {
      img.addEventListener('error', convertedProps.onError);
    }
    if (convertedProps.crossOrigin) {
      img.crossOrigin = convertedProps.crossOrigin;
    }

    // Only directly activate the image if critical (preload).
    if ((convertedProps.critical || convertedProps.isVisible) && !isLoop) {
      return activatePictureRef(img, convertedProps, index, isLoop);
    }

    return img;
  }
  return null;
};

/**
 * Creates multiple image references. Internal function.
 *
 * @param props   object    Component Props (with fluid or fixed as array).
 * @param onLoad  function  Callback for load handling.
 */
export const createMultiplePictureRefs = (props, onLoad) => {
  const convertedProps = convertProps(props);

  // Extract Image Array.
  const imageStack = convertedProps.fluid || convertedProps.fixed;
  const imageRefs = imageStack.map((imageData, index) =>
    createPictureRef(convertedProps, onLoad, index, true)
  );
  // Only directly activate the image if critical (preload).
  if (convertedProps.critical || convertedProps.isVisible) {
    return activatePictureRef(imageRefs, convertedProps);
  }
  return imageRefs;
};

/**
 * Creates a picture element for the browser to decide which image to load.
 *
 * @param imageRef
 * @param props
 * @param selfRef
 * @param index
 * @param isLoop
 * @return {null|Array|*}
 */
export const activatePictureRef = (
  imageRef,
  props,
  selfRef = null,
  index = 0,
  isLoop = false
) => {
  const convertedProps = convertProps(props);
  if (
    isBrowser() &&
    (typeof convertedProps.fluid !== `undefined` ||
      typeof convertedProps.fixed !== `undefined`)
  ) {
    const isImageStack =
      hasImageArray(convertedProps) && !hasArtDirectionArray(convertedProps);
    if (isImageStack && !isLoop) {
      return activateMultiplePictureRefs(imageRef, props, selfRef);
    }
    // Clone body to get the correct sizes.
    // const bodyClone = document.body.cloneNode(true)
    const bodyClone = document.createElement('body');
    // Do we have an image stack or Art-direction array?
    // Then get its current or first(smallest) image respectively.
    const imageData = isImageStack
      ? getSelectedImage(convertedProps, index)
      : getCurrentSrcData(convertedProps);

    if (!imageData) {
      return null;
    }
    if (isString(imageData)) {
      return imageData;
    }

    if (selfRef) {
      // Set original component's style.
      imageRef.width = selfRef.offsetWidth;
      imageRef.height = selfRef.offsetHeight;
    }

    // Prevent adding HTMLPictureElement if it isn't supported (e.g. IE11),
    // but don't prevent it during SSR.
    if (hasPictureElement()) {
      const pic = document.createElement('picture');
      if (selfRef) {
        // Set original component's style.
        pic.width = imageRef.width;
        pic.height = imageRef.height;
      }
      if (hasArtDirectionArray(convertedProps)) {
        const sources = createArtDirectionSources(convertedProps).reverse();
        sources.forEach(currentSource => pic.appendChild(currentSource));
      }

      const sourcesAvif = createSourceElementForSrcSet(imageData, 'avif');
      sourcesAvif && pic.appendChild(sourcesAvif);
      const sourcesWebp = createSourceElementForSrcSet(imageData, 'webp');
      sourcesWebp && pic.appendChild(sourcesWebp);

      pic.appendChild(imageRef);
      bodyClone.appendChild(pic);
    }

    imageRef.sizes = imageData.sizes || ``;
    imageRef.srcset = imageData.srcSet || ``;
    imageRef.src = imageData.src || ``;

    return imageRef;
  }
  return null;
};

/**
 * Creates multiple picture elements.
 *
 * @param imageRefs
 * @param props
 * @param selfRef
 * @return {Array||null}
 */
export const activateMultiplePictureRefs = (imageRefs, props, selfRef) => {
  // Extract Image Array.
  return imageRefs.map((imageRef, index) =>
    activatePictureRef(imageRef, props, selfRef, index, true)
  );
};

/**
 * Checks imageRefs on being active.
 *
 * @param imageRefs
 * @return {*}
 */
export const hasActivatedPictureRefs = imageRefs => {
  return Array.isArray(imageRefs)
    ? imageRefs.every(imageRef => hasPictureRef(imageRef))
    : hasPictureRef(imageRefs);
};

/**
 * Checks imageRef for on being a string or has a currentSrc.
 *
 * @param imageRef
 * @return {boolean}
 */
export const hasPictureRef = imageRef => {
  return isString(imageRef) || (!!imageRef && !!imageRef.currentSrc);
};

/**
 * Checks if an image (array) reference is existing and tests for complete.
 *
 * @param imageRef    HTMLImageElement||array   Image reference(s).
 * @param props
 * @return {boolean}
 */
export const imageReferenceCompleted = (imageRef, props) =>
  imageRef
    ? Array.isArray(imageRef)
      ? imageRef.every(singleImageRef => imageLoaded(singleImageRef)) ||
        inImageCache(props)
      : imageLoaded(imageRef) || inImageCache(props)
    : isString(imageRef);
