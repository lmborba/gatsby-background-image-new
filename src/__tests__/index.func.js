import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import './mocks/IntersectionObserver.mock';
import { mockAllIsIntersecting } from './mocks/IntersectionObserver.mock';

import {
  fixedShapeMock,
  fluidShapeMock,
  setupBackgroundImage,
} from './mocks/Various.mock';
import BackgroundImage from '..';
import { activateCacheForImage, resetImageCache } from '../lib/ImageCache';
import { resetComponentClassCache } from '../lib/ClassCache';
import { createPictureRef } from '../lib/ImageRef';

const LOAD_FAILURE_SRC = 'test_fluid_image.jpg';
const LOAD_SUCCESS_SRC = 'test_fixed_image.jpg';

describe(`<BackgroundImage /> with mock IO`, () => {
  const tmpImagePrototype = Object.getPrototypeOf(HTMLImageElement);

  beforeEach(() => {
    // Freeze StyleUtils#fixClassName.
    const uuid = require('short-uuid');
    uuid.generate.mockImplementation(() => '73WakrfVbNJBaAmhQtEeDv');

    // Mocking HTMLImageElement.prototype.src to call the onload or onerror
    // callbacks depending on the src passed to it
    Object.defineProperty(HTMLImageElement.prototype, 'src', {
      // Define the property setter
      set(src) {
        this.setAttribute('src', src);
        if (src === LOAD_FAILURE_SRC) {
          // Call with setTimeout to simulate async loading
          try {
            const errorEvent = new Event('error');
            setTimeout(() => this.dispatchEvent(errorEvent));
          } catch (e) {
            console.log(src, e);
          }
        } else if (src === LOAD_SUCCESS_SRC) {
          // Call with setTimeout to simulate async loading
          try {
            const errorEvent = new Event('load');
            setTimeout(() => this.dispatchEvent(errorEvent));
          } catch (e) {
            console.log(src, e);
          }
        }
      },
    });
    Object.defineProperty(HTMLImageElement.prototype, 'complete', {
      // Define the property setter
      set(complete) {
        this.complete = complete;
      },
      get() {
        if (this.getAttribute('src') === LOAD_SUCCESS_SRC) {
          return true;
        }
      },
    });
    Object.defineProperty(HTMLImageElement.prototype, 'naturalWidth', {
      // Define the property setter
      set(naturalWidth) {
        this.naturalWidth = naturalWidth;
      },
      get() {
        if (this.getAttribute('src') === LOAD_SUCCESS_SRC) {
          return 1;
        }
      },
    });
    Object.defineProperty(HTMLImageElement.prototype, 'naturalHeight', {
      // Define the property setter
      set(naturalHeight) {
        this.naturalHeight = naturalHeight;
      },
      get() {
        if (this.getAttribute('src') === LOAD_SUCCESS_SRC) {
          return 1;
        }
      },
    });

    resetImageCache();
  });
  afterEach(() => {
    Object.setPrototypeOf(HTMLImageElement, tmpImagePrototype);
    resetComponentClassCache();
  });

  it(`should render visible fluid image`, () => {
    // Mock Math.random beforehand, lest another random classname is created.
    Math.random = jest.fn(() => 0.424303425546642);
    let { container, rerender } = render(
      <BackgroundImage fluid={fluidShapeMock} />
    );
    mockAllIsIntersecting(true);
    expect(container).toMatchSnapshot();
    container = rerender(<BackgroundImage fluid={fluidShapeMock} />);
    expect(container).toMatchSnapshot();
  });

  it(`should render visible fixed image and call onLoadFunction`, () => {
    const currentFluidMock = {
      src: ``,
      aspectRatio: 1,
      srcSet: ``,
      sizes: ``,
    };
    // Mock Math.random beforehand, lest another random classname is created.
    Math.random = jest.fn(() => 0.424303425546642);
    // Mock onStartLoad().
    const onStartLoadFunctionMock = jest.fn();
    let { container, rerender } = render(
      <BackgroundImage
        fluid={currentFluidMock}
        onStartLoad={onStartLoadFunctionMock}
      />
    );
    mockAllIsIntersecting(true);
    expect(container).toMatchSnapshot();
    expect(onStartLoadFunctionMock).toHaveBeenCalled();
    container = rerender(<BackgroundImage fixed={fixedShapeMock} />);
    expect(container).toMatchSnapshot();
  });

  it(`should call critical fixed images`, () => {
    activateCacheForImage({ fixed: fixedShapeMock });
    const options = {
      addClass: true,
      critical: true,
      fixed: true,
    };
    const component = setupBackgroundImage(options);
    mockAllIsIntersecting(true);
    expect(component).toMatchSnapshot();
  });

  it(`should call stacked critical fixed images`, () => {
    activateCacheForImage({ fixed: [fixedShapeMock, fixedShapeMock] });
    const options = {
      addClass: true,
      critical: true,
      fixed: true,
      multiImage: true,
    };
    const component = setupBackgroundImage(options);
    mockAllIsIntersecting(true);
    expect(component).toMatchSnapshot();
  });

  it(`should call stacked critical fluid images`, () => {
    activateCacheForImage({ fluid: [fluidShapeMock, fluidShapeMock] });
    const options = {
      addClass: true,
      critical: true,
      fixed: false,
      fluid: true,
      multiImage: true,
    };
    const component = setupBackgroundImage(options);
    mockAllIsIntersecting(true);
    expect(component).toMatchSnapshot();
  });

  it(`should not call onLoad without prop, fadeIn should stay falsy without seenBefore`, () => {
    const options = {
      addClass: true,
      critical: true,
      fixed: true,
      onLoad: null,
    };
    const component = setupBackgroundImage(options);
    mockAllIsIntersecting(true);
    expect(component).toMatchSnapshot();
  });

  it(`should call onLoad and onError image events for LOAD_SUCCESS_SRC`, () => {
    const onLoadMock = jest.fn();
    const onErrorMock = jest.fn();
    const image = createPictureRef(
      {
        fixed: {
          src: LOAD_SUCCESS_SRC,
        },
        onLoad: onLoadMock,
        onError: onErrorMock,
      },
      onLoadMock
    );
    fireEvent.load(image);
    fireEvent.error(image);

    expect(onLoadMock).toHaveBeenCalledTimes(2);
    expect(onErrorMock).toHaveBeenCalledTimes(1);
  });

  it(`shouldn't call onLoad and onError image events without src`, () => {
    const onLoadMock = jest.fn();
    const onErrorMock = jest.fn();
    const image = createPictureRef(
      {
        fluid: {
          src: ``,
        },
        onLoad: onLoadMock,
        onError: onErrorMock,
      },
      onLoadMock
    );
    fireEvent.load(image);
    fireEvent.error(image);

    expect(onLoadMock).toHaveBeenCalledTimes(2);
    expect(onErrorMock).toHaveBeenCalledTimes(1);
  });
});

describe(`<BackgroundImage /> without IO`, () => {
  const tmpIO = global.IntersectionObserver;
  beforeEach(() => {
    delete global.IntersectionObserver;
    resetImageCache();
  });
  afterEach(() => {
    global.IntersectionObserver = tmpIO;
  });

  it(`should call onLoadFunction without IO`, () => {
    const onLoadFunctionMock = jest.fn();
    activateCacheForImage({ fluid: fluidShapeMock });
    const options = {
      fluid: true,
      addClass: true,
      additionalClass: `test`,
      critical: true,
      onStartLoad: onLoadFunctionMock,
      fixed: false,
    };
    const component = setupBackgroundImage(options);
    expect(component).toMatchSnapshot();
    expect(onLoadFunctionMock).toHaveBeenCalled();
  });
});
