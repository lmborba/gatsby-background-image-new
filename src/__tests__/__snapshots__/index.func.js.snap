// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BackgroundImage /> with mock IO should call critical fixed images 1`] = `
<div>
  <div
    alt="Alt text for the image"
    class="imageClass  gbi-1393017994-73WakrfVbNJBaAmhQtEeDv"
    id="testid"
    style="background-position: center; background-repeat: repeat-y; background-size: contain; position: relative; display: inline-block; opacity: 0.99; width: 100px; height: 100px;"
    title="Title for the image"
  >
    <style>
      
          .imageClass.gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:before,
          .imageClass.gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:after {
            content: '';
            display: block;
            position: absolute;
            width: 100%;
            height: 100%;
            top: 0;
            left: 0;
            background-color: lightgray;
            transition: none;
            background-position: center;
background-repeat: repeat-y;
background-size: contain;

          }
          .imageClass.gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:before {
            z-index: -100;
            background-image: url('http://localhost/test_fixed_image.jpg');
            background-image: url(string_of_base64);
            opacity: 0; 
          }
          .imageClass.gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:after {
            z-index: -101;
            background-image: url('http://localhost/test_fixed_image.jpg');
            
            opacity: 1;
          }
        
    </style>
    <h1>
      test
    </h1>
  </div>
</div>
`;

exports[`<BackgroundImage /> with mock IO should call stacked critical fixed images 1`] = `
<div>
  <div
    alt="Alt text for the image"
    class="imageClass  gbi-1393017994-73WakrfVbNJBaAmhQtEeDv"
    id="testid"
    style="background-position: center; background-repeat: repeat-y; background-size: contain; position: relative; display: inline-block; opacity: 0.99;"
    title="Title for the image"
  >
    <style>
      
          .imageClass.gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:before,
          .imageClass.gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:after {
            content: '';
            display: block;
            position: absolute;
            width: 100%;
            height: 100%;
            top: 0;
            left: 0;
            background-color: lightgray;
            transition: none;
            background-position: center;
background-repeat: repeat-y;
background-size: contain;

          }
          .imageClass.gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:before {
            z-index: -100;
            background-image: url('http://localhost/test_fixed_image.jpg'),url('http://localhost/test_fixed_image.jpg');
            background-image: url(string_of_base64),url(string_of_base64);
            opacity: 0; 
          }
          .imageClass.gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:after {
            z-index: -101;
            background-image: url('http://localhost/test_fixed_image.jpg'),url('http://localhost/test_fixed_image.jpg');
            
            opacity: 1;
          }
        
    </style>
    <h1>
      test
    </h1>
  </div>
</div>
`;

exports[`<BackgroundImage /> with mock IO should call stacked critical fluid images 1`] = `
<div>
  <div
    alt="Alt text for the image"
    class="imageClass  gbi-1393017994-73WakrfVbNJBaAmhQtEeDv"
    id="testid"
    style="background-position: center; background-repeat: repeat-y; background-size: contain; position: relative; display: inline; opacity: 0.99;"
    title="Title for the image"
  >
    <style>
      
          .imageClass.gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:before,
          .imageClass.gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:after {
            content: '';
            display: block;
            position: absolute;
            width: 100%;
            height: 100%;
            top: 0;
            left: 0;
            background-color: lightgray;
            transition: opacity 0.5s ease 250ms;
            background-position: center;
background-repeat: repeat-y;
background-size: contain;

          }
          .imageClass.gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:before {
            z-index: -100;
            background-image: url(string_of_base64),url(string_of_base64),linear-gradient(to right, rgba(30, 75, 115, 1), rgba(255, 255, 255, 0));
            
            opacity: 1; 
          }
          .imageClass.gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:after {
            z-index: -101;
            
            background-image: url(string_of_base64),url(string_of_base64),linear-gradient(to right, rgba(30, 75, 115, 1), rgba(255, 255, 255, 0));
            
          }
        
    </style>
    <h1>
      test
    </h1>
  </div>
</div>
`;

exports[`<BackgroundImage /> with mock IO should not call onLoad without prop, fadeIn should stay falsy without seenBefore 1`] = `
<div>
  <div
    alt="Alt text for the image"
    class="imageClass  gbi-1393017994-73WakrfVbNJBaAmhQtEeDv"
    id="testid"
    style="background-position: center; background-repeat: repeat-y; background-size: contain; position: relative; display: inline-block; opacity: 0.99; width: 100px; height: 100px;"
    title="Title for the image"
  >
    <style>
      
          .imageClass.gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:before,
          .imageClass.gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:after {
            content: '';
            display: block;
            position: absolute;
            width: 100%;
            height: 100%;
            top: 0;
            left: 0;
            background-color: lightgray;
            transition: opacity 0.5s ease 250ms;
            background-position: center;
background-repeat: repeat-y;
background-size: contain;

          }
          .imageClass.gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:before {
            z-index: -100;
            background-image: url('http://localhost/test_fixed_image.jpg');
            background-image: url(string_of_base64);
            opacity: 0; 
          }
          .imageClass.gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:after {
            z-index: -101;
            background-image: url('http://localhost/test_fixed_image.jpg');
            
            opacity: 1;
          }
        
    </style>
    <h1>
      test
    </h1>
  </div>
</div>
`;

exports[`<BackgroundImage /> with mock IO should render visible fixed image and call onLoadFunction 1`] = `
<div>
  <div
    class="gbi-2113735767-73WakrfVbNJBaAmhQtEeDv"
    style="background-position: center; background-repeat: no-repeat; background-size: cover; position: relative; opacity: 0.99;"
  >
    <style>
      
          .gbi-2113735767-73WakrfVbNJBaAmhQtEeDv:before,
          .gbi-2113735767-73WakrfVbNJBaAmhQtEeDv:after {
            content: '';
            display: block;
            position: absolute;
            width: 100%;
            height: 100%;
            top: 0;
            left: 0;
            
            transition: opacity 0.5s ease 250ms;
            background-position: center;
background-repeat: no-repeat;
background-size: cover;

          }
          .gbi-2113735767-73WakrfVbNJBaAmhQtEeDv:before {
            z-index: -100;
            
            
            opacity: 1; 
          }
          .gbi-2113735767-73WakrfVbNJBaAmhQtEeDv:after {
            z-index: -101;
            
            
            
          }
        
    </style>
  </div>
</div>
`;

exports[`<BackgroundImage /> with mock IO should render visible fixed image and call onLoadFunction 2`] = `undefined`;

exports[`<BackgroundImage /> with mock IO should render visible fluid image 1`] = `
<div>
  <div
    class="gbi-1393017994-73WakrfVbNJBaAmhQtEeDv"
    style="background-position: center; background-repeat: no-repeat; background-size: cover; position: relative; opacity: 0.99;"
  >
    <style>
      
          .gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:before,
          .gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:after {
            content: '';
            display: block;
            position: absolute;
            width: 100%;
            height: 100%;
            top: 0;
            left: 0;
            
            transition: opacity 0.5s ease 250ms;
            background-position: center;
background-repeat: no-repeat;
background-size: cover;

          }
          .gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:before {
            z-index: -100;
            background-image: url(string_of_base64);
            
            opacity: 1; 
          }
          .gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:after {
            z-index: -101;
            
            background-image: url(string_of_base64);
            
          }
        
    </style>
  </div>
</div>
`;

exports[`<BackgroundImage /> with mock IO should render visible fluid image 2`] = `undefined`;

exports[`<BackgroundImage /> without IO should call onLoadFunction without IO 1`] = `
<div>
  <div
    alt="Alt text for the image"
    class="imageClass test gbi-1393017994-73WakrfVbNJBaAmhQtEeDv"
    id="testid"
    style="background-position: center; background-repeat: repeat-y; background-size: contain; position: relative; display: inline; opacity: 0.99;"
    title="Title for the image"
  >
    <style>
      
          .imageClass.test.gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:before,
          .imageClass.test.gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:after {
            content: '';
            display: block;
            position: absolute;
            width: 100%;
            height: 100%;
            top: 0;
            left: 0;
            background-color: lightgray;
            transition: none;
            background-position: center;
background-repeat: repeat-y;
background-size: contain;

          }
          .imageClass.test.gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:before {
            z-index: -100;
            background-image: url('http://localhost/test_fluid_image.jpg');
            background-image: url(string_of_base64);
            opacity: 0; 
          }
          .imageClass.test.gbi-1393017994-73WakrfVbNJBaAmhQtEeDv:after {
            z-index: -101;
            background-image: url('http://localhost/test_fluid_image.jpg');
            
            opacity: 1;
          }
        
    </style>
    <h1>
      test
    </h1>
  </div>
</div>
`;
